### ISSUE TITLE
OpenXR Motion Compensation causes freeze and very low fps with VDXR

### ISSUE THREAD

**Issue by pdebaty** [mbucchia/VirtualDesktop-OpenXR#36]
**Describe the bug**

When using OpenXR Motion Compensation (OXMRC) with a motion platform (DOF reality H2) and running a game like DCS or MSFS (same issue with simracing as well) and when doing a sharp maneuver (which should cause a big compensation from OXRMC), the picture freezes for 2-3 secs and afterwards its running at 10-20 fps with lots of studdering. At that point, if I toggle VR off and on again, the normal framerate is back and everything works fine until the next sharp maneuver.
This issue does not occur when using another OpenXR system like Steam or Oculus. It only occurs with VDXR.

I am attaching the debug trace of the issue:
with OpenXR toolkit and turbo mode enabled: [VDXR.zip](https://github.com/mbucchia/VirtualDesktop-OpenXR/files/14390379/VDXR.zip)
With OpenXR toolkit disabled: 
[VDXR (2).zip](https://github.com/mbucchia/VirtualDesktop-OpenXR/files/14390674/VDXR.2.zip)

Also attaching my OpenXR.log file: 
[OpenXR.log](https://github.com/mbucchia/VirtualDesktop-OpenXR/files/14390586/OpenXR.log)

Here is a video of another user having the same issue:  https://youtube.com/shorts/ZqxZtpncIf8?feature=share

**Environment**

- _Application/Game and version_: DCS or MSFS, or any other flight sim or sim racing game that uses motion compensation
- _Version of Virtual Desktop_: 1.30.1

**To Reproduce**

_Steps to reproduce the behavior:_

1. Run DCS or MSFS or other flight/simracing game.
2. Enable OpenXR Motion Compensation.
3. Perform a sharp maneuver in game (quick roll, or crash the plane or car)
4. Observe a freeze/black screen followed by low fps, jitters and stutters.

**Expected behavior**

No freeze and no jitters/stutters afterwards.

**Comment by pdebaty** [comment]
debug trace with OpenKneeboard turned off: 
[VDXR (3).zip](https://github.com/mbucchia/VirtualDesktop-OpenXR/files/14390809/VDXR.3.zip)

**Comment by pdebaty** [comment]
debug trace with HTCC and OpenKneeboard disabled:
[VDXR (4).zip](https://github.com/mbucchia/VirtualDesktop-OpenXR/files/14390866/VDXR.4.zip)

**Comment by mbucchia** [comment]
I'm adding a (hidden) setting for tuning thread priorities, just in case something of the sort is happening. I'll share how to tune it once the release is out.

112a6aa1bbafbabb0d912b78d5f13617925fb3b3

**Comment by BuzzteeBear** [comment]
Preliminary result of my investigation seem to indicate that the issue seems to be related to the eye poses reported to the runtime in xrEndFrame. 
Background:
OXRMC needs to manipulate the eye poses coming from the application in xrEndFrame as it manipulated the poses it previously reported to the application via xrLocateViews, so they don't match the actual real hmd pose.
Currently there are two different methods available how oxrmc obtains these 'reconstructed' eye poses:
1. Default: invert the pose that was used to manipulate the eye poses in XrLocateViews and apply them to the poses reported by the application in xrEndFrame
2. Save the original eye poses supplied by the runtime in xrLocateViews and reimplant them on xrEndFrame

The issue only seems to occur oonly when using the 1. method. Method 2. seems to be fine in that regard, but has other issues e.g. introduces some camera jitter in MSFS.
The user can switch between both methods using ctrl+e in game or setting the flag `use_eye_cache` in oxrmc's .ini file

Interestingly, if the issue started to happen partt of it seems to persist even after switching to the other method or even when deactivating motion compensation via ctrl+insert

**Comment by romandesign** [comment]
It's worth adding that it only happens with SSW DISABLED, and cannot be reproduced if SSW is enabled.

**Comment by BuzzteeBear** [comment]
Just quick heads up so you don't have to waste too much time on that issue:

After giving some more thought into the things I stated in my previous comnment I had a hunch that the issue might be caused by too much calculation error. So I decided to normalize the quaternions within the eye poses before submitting them in xrEndFrame. At least on my system that seems to have elimnated the issue in the scenarios where I could reliably cause it.
I gave out a quick and dirty version of the fix to have it confirmed by other users as well and the verdict on that is still to come.

So long story short: it seems oxrmc just fed non-normalized quaternions to the runtime.

I can't tell why VDXR seems to be more sensitive to that fact than other runtimes but that might be worthwile to investigate.
 
If the feedback on my preliminary fix is positive, I'll make sure to go through all poses oxrmc is touching and normalize their quaternions where it seems appropriate

**Comment by mbucchia** [comment]
Interesting find. I don't know/think there is a constraint in the API where quaternions need to be normalized. I certain don't have the extra call to do it on VDXR, built I could add it.

Unclear why it would manifest in a beginFrame timeout though...

**Comment by mbucchia-msft** [comment]
OK well this is weird, never paid attention to it:

https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#coordinate-system

> Rotation is represented by a **unit quaternion** defined by the [XrQuaternionf](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XrQuaternionf) structure:

> A runtime must return XR_ERROR_POSE_INVALID if the orientation norm deviates by more than 1% from unit **length.**

I guess we are all doing it wrong lol. Runtimes should be rejecting your quaternions with an error, but they probably just always normalize them instead.

🤷

**Comment by mbucchia-msft** [comment]
I'm even more puzzled now, because VDXR does check and error-out on non-normalized quaternions: https://github.com/mbucchia/VirtualDesktop-OpenXR/blob/main/virtualdesktop-openxr/frame.cpp#L462

So I'm really not sure what we are looking at here(?)

**Comment by BuzzteeBear** [comment]
Looks like some computation on the pose is done further down: https://github.com/mbucchia/VirtualDesktop-OpenXR/blob/0d2a45704aa847f62d819cdbc546237b7c235ccb/virtualdesktop-openxr/frame.cpp#L507

So maybe the quaternions coming in are just on the edge of not being normalized and that last multiplication(s) are what breaks the the camels back.

But that's pure speculation on my part...
